### PR TITLE
[KED-1878] Add Avro dataset

### PR DESCRIPTION
**Pull request by kislerdm**
## Description

The PR concludes the issue #607: 
- Adds support of avro dataset
- Adds support to unload avro datasets from GBQ to GCS
- Adds support for pandas DataFrames interface to AVRO format

## Development notes

Test using unit test with pytest and by mimicking a real case scenario through loading/exporting a table to/from GBQ from/to a GCS bucket.

sphinx seems to fail to figure the link between the pandas avro class and the base avro class it inherits methods from, i.e.

```bash
make build-docs
```
throws a warning which prevents documentation from compiling. Why the shynx build parameter `-W` is being used?

## Checklist

- [x] Read the [contributing](https://github.com/quantumblacklabs/kedro/blob/master/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](https://github.com/quantumblacklabs/kedro/blob/master/RELEASE.md) file
- [x] Added tests to cover my changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

**Comment by yetudada** [comment]
This is fantastic @kislerdm! We'll put it on the backlog to review!

File: kedro/extras/datasets/avro/gbq_to_gcs_dataset.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+        self._client = bigquery.Client(
+            project=project, credentials=credentials, location=location,
+        )
+
+        self._table_id = f"{project}.{dataset}.{table_name}"
+        self._table_ref = self._client.dataset(dataset).table(table_name)
+
+        path_object = path_object or f"{dataset}/{table_name}.avro"
+        self._uri = f"""gs://{bucket}/{path_object}"""
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/avro/gbq_to_gcs_dataset.py:160]:
```suggestion
        self._uri = f"gs://{bucket}/{path_object}"
```

File: requirements.txt [hunk 2]
```diff
@@ -1,4 +1,5 @@
 anyconfig~=0.9.7
+avro==1.10.0
```
Comment (MerelTheisenQB) [inline requirements.txt:2]: Can you move this to `test_requirements.txt` instead? Avro isn't needed to use the core functionality of Kedro. And would it be possible to pin it more loosely instead of strictly to 1.10.0?

File: kedro/extras/datasets/avro/avro_dataset.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+"""``AVRODataSet`` loads/saves data from/to a AVRO file using an underlying
+filesystem (e.g.: local, S3, GCS). It uses avro library to handle the AVRO file.
+See details about AVRO format: https://avro.apache.org/
+"""
+import json
+from copy import deepcopy
+from pathlib import Path, PurePosixPath
+from typing import Any, Dict, List, Optional
+
+import fsspec  # type: ignore
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/avro/avro_dataset.py:38]: What is the `# type: ignore` for here and for the other imports?

File: kedro/extras/datasets/avro/avro_dataset.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                Here you can find all available arguments for `open`:
+                https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem.open
+                All defaults are preserved, except `mode`, which is set to `rb` when loading
+                and to `wb` when saving.
+        """
+        _credentials = deepcopy(credentials) or {}
+        _fs_args = deepcopy(fs_args) or {}
+
+        self._fs_open_args_load = _fs_args.pop("open_args_load", {})
+        self._fs_open_args_load.setdefault("mode", "rb")
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/avro/avro_dataset.py:132]: `rb` is already the default value, so no need to set it here.

File: kedro/extras/datasets/avro/gbq_to_gcs_dataset.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
+# Trademarks or any confusingly similar mark as a trademark for your product,
+# or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""``GBQTableGCSAVRODataSet`` unload/load data from/to a GBQ table as/using an AVRO file
```
Comment (MerelTheisenQB) [inline kedro/extras/datasets/avro/gbq_to_gcs_dataset.py:29]: I had to read this several times to make sense of it. Could you rephrase this to make it a bit clearer what this dataset is for exactly?

File: tests/extras/datasets/avro/test_avro_dataset.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    def test_credentials_propagated(self, mocker):
+        """Test propagating credentials for connecting to GCS"""
+        mock_fs = mocker.patch("fsspec.filesystem")
+        credentials = {"key": "value"}
+
+        AVRODataSet(filepath=FILENAME, credentials=credentials)
+
+        mock_fs.assert_called_once_with("file", **credentials)
+
+    def test_save_and_load(self, tmp_path, dummy_data, dummy_schema):
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/avro/test_avro_dataset.py:122]: Could you divide this up in several tests? Ideally a unit test would only cover 1 case, so just one happy path or one unhappy path, not a combination of several.

File: tests/extras/datasets/avro/test_avro_dataset.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        """Test `exists` method invocation for both existing and
+        nonexistent data set."""
+        assert not avro_data_set.exists()
+        avro_data_set.save(dummy_data)
+        assert avro_data_set.exists()
+
+        def patch_error():
+            raise DataSetError("test")
+
+        avro_data_set._get_load_path = patch_error
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/avro/test_avro_dataset.py:164]: It's not very clear to me what you're testing here. Could you move it to a separate test with an explanatory name?

File: tests/extras/datasets/avro/test_gbq_to_gcs_dataset.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    def test_exists(self, mock_bigquery_client):
+        """Test `exists` method invocation."""
+        mock_bigquery_client.return_value.get_table.side_effect = [
+            NotFound("NotFound"),
+            "exists",
+        ]
+
+        data_set = GBQTableGCSAVRODataSet(PROJECT, DATASET, TABLE_NAME, BUCKET)
+        assert not data_set.exists()
+        assert data_set.exists()
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/avro/test_gbq_to_gcs_dataset.py:78]: These asserts are the opposite so how can they both pass in the same test?

File: tests/extras/datasets/avro/test_gbq_to_gcs_dataset.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        for key, value in save_args.items():
+            assert gbq_to_gcs_avro_dataset._save_args[key] == value
+
+    def test_load(self, gbq_to_gcs_avro_dataset):
+        assert not gbq_to_gcs_avro_dataset.load()
+
+    def test_save(self, gbq_to_gcs_avro_dataset):
+        assert not gbq_to_gcs_avro_dataset.save()
+
+    def test_save_exceptions(self, gbq_to_gcs_avro_dataset):
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/avro/test_gbq_to_gcs_dataset.py:102]: Can you divide this up into two tests each with a descriptive name of the exception that's expected to happen?

File: tests/extras/datasets/pandas/test_avro_dataset.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    def test_credentials_propagated(self, mocker):
+        """Test propagating credentials for connecting to GCS"""
+        mock_fs = mocker.patch("fsspec.filesystem")
+        credentials = {"key": "value"}
+
+        AVRODataSet(filepath=FILENAME, credentials=credentials)
+
+        mock_fs.assert_called_once_with("file", **credentials)
+
+    def test_save_and_load(self, tmp_path, dummy_data):
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/pandas/test_avro_dataset.py:73]: Can you divide this up into two tests, one for the happy path and one for the unhappy path?

File: tests/extras/datasets/pandas/test_avro_dataset.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        nonexistent data set."""
+        assert not avro_pd_data_set.exists()
+        avro_pd_data_set.save(dummy_data)
+        assert avro_pd_data_set.exists()
+
+        def patch_error():
+            raise DataSetError("test")
+
+        avro_pd_data_set._get_load_path = patch_error
+        assert not avro_pd_data_set.exists()
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/pandas/test_avro_dataset.py:97]: Similar comment to the other test_avro_dataset, it's not very clear what you're testing here so could you move it to a separate test with an explanatory name?

File: tests/extras/datasets/pandas/test_avro_dataset.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+                },
+            },
+            {"in": numpy.nan, "want": {"type": ["null", "string"]}},
+        ]
+
+        for test in tests:
+            got = infer_pandas_dtype_to_avro_type(test["in"])
+            assert got == test["want"]
+
+    def test_unhappy_path(self):
```
Comment (MerelTheisenQB) [inline tests/extras/datasets/pandas/test_avro_dataset.py:189]: Can you make this name more descriptive, it's very generic now.

**Review by MerelTheisenQB — COMMENTED** [review]
Thank you very much for contributing this @kislerdm ! I've added some questions and comments.

File: requirements.txt (hunk 2 above)
Comment (kislerdm) [reply requirements.txt:2]: What is the purpose of the `test_requirements.txt` file?

File: docs/build-docs.sh [hunk 13]
```diff
@@ -43,7 +43,7 @@ rm -rf docs/build
 mkdir docs/build/
 cp -r docs/_templates docs/conf.py docs/*.svg docs/*.json  docs/build/
 
-sphinx-build -c docs/ -WETan -j auto -D language=en docs/build/ docs/build/html
+sphinx-build -c docs/ -ETan -j auto -D language=en docs/build/ docs/build/html
```
[the remark this answers is not in the document]
Comment (kislerdm) [reply docs/build-docs.sh:46]: Hey! It passes sphinx execution through in case of warnings. In the cast of the flag `-W`, a sphinx process is being terminated on warning. 
The proposal to remove the flag was motivated by the issue it causes when running sphinx to generate docu for [avro_dataset.py](https://github.com/kislerdm/kedro/blob/feature/issue-607_adds_avro_dataset/kedro/extras/datasets/pandas/avro_dataset.py).

File: kedro/extras/datasets/avro/avro_dataset.py (hunk 3 above)
Comment (kislerdm) [reply kedro/extras/datasets/avro/avro_dataset.py:38]: `mypy` was throwing errors in my venv, it is to skips mypy checks of that lines

File: kedro/extras/datasets/avro/avro_dataset.py (hunk 4 above)
Comment (kislerdm) [reply kedro/extras/datasets/avro/avro_dataset.py:132]: I'd rather leave it as an explicitly specified attribute, what do you think?

File: docs/source/14_api_docs/kedro.extras.datasets.rst [hunk 14]
```diff
@@ -13,12 +13,15 @@ Data Sets
     :template: autosummary/class.rst
 
     kedro.extras.datasets.api.APIDataSet
+    kedro.extras.datasets.avro.AVRODataSet
```
Comment (deepyaman) [inline docs/source/14_api_docs/kedro.extras.datasets.rst:16]: Avro shouldn't be all capitalized, as it's not an abbreviation (here or anywhere after).

```suggestion
    kedro.extras.datasets.avro.AvroDataSet
```

File: requirements.txt (hunk 2 above)
Comment (deepyaman) [reply requirements.txt:2]: @kislerdm `test_requirements.txt` is used to specify an extended set of dependencies required to run the full unit test suite, among other things. If you look at [the first line](https://github.com/quantumblacklabs/kedro/blob/master/test_requirements.txt#L1), you'll see that it includes all of `requirements.txt`.

While Avro isn't core to the framework, you'll also need to add lines in [setup.py](https://github.com/quantumblacklabs/kedro/blob/master/setup.py) to include the appropriate Avro dependencies when somebody wants to use the dataset.

File: docs/build-docs.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (MerelTheisenQB) [reply docs/build-docs.sh:46]: Ah I see, thanks for explaining. We do actually want to keep the `-W` flag, because for our building of the docs the process should terminate on warning and not continue.

File: kedro/extras/datasets/avro/avro_dataset.py (hunk 3 above)
Comment (MerelTheisenQB) [reply kedro/extras/datasets/avro/avro_dataset.py:38]: Okay, thanks for clarifying! We're not actually using `mypy` as a static check for Kedro, so we'd prefer not to have those ignore statements in the code.

File: kedro/extras/datasets/avro/avro_dataset.py (hunk 4 above)
Comment (MerelTheisenQB) [reply kedro/extras/datasets/avro/avro_dataset.py:132]: In all the other datasets we don't explicitly define the load mode if it's using the default, so I'd prefer to keep it consistent and remove it here as well.

**Comment by MerelTheisenQB** [comment]
Hi @kislerdm, will you be able to finish this PR or are you happy with me jumping in and making it ready for merging?
